python/ray/dashboard/memory.py

```diff
@@ -0,0 +1,280 @@
+import base64
+import typing
+
+from collections import defaultdict
+from enum import Enum
+
+import ray
+
+
+def decode_object_id_if_needed(object_id: str) -> bytes:
+    """Decode objectID bytes string.
+
+    gRPC reply contains an objectID that is encodded by Base64.
+    This function is used to decode the objectID.
+    Note that there are times that objectID is already decoded as
+    a hex string. In this case, just convert it to a binary number.
+    """
+    if object_id.endswith("="):
+        # If the object id ends with =, that means it is base64 encoded.
+        # Object ids will always have = as a padding
+        # when it is base64 encoded because objectID is always 20B.
+        return base64.standard_b64decode(object_id)
+    else:
+        return ray.utils.hex_to_binary(object_id)
+
+
+class SortingType(Enum):
+    PID = 1
+    OBJECT_SIZE = 3
+    REFERENCE_TYPE = 4
+
+
+class GroupByType(Enum):
+    NODE_ADDRESS = 2
+
+
+class ReferenceType:
+    # We don't use enum because enum is not json serializable.
+    ACTOR_HANDLE = "ACTOR_HANDLE"
+    PINNED_IN_MEMORY = "PINNED_IN_MEMORY"
+    LOCAL_REFERENCE = "LOCAL_REFERENCE"
+    USED_BY_PENDING_TASK = "USED_BY_PENDING_TASK"
+    CAPTURED_IN_OBJECT = "CAPTURED_IN_OBJECT"
+    UNKNOWN_STATUS = "UNKNOWN_STATUS"
+
+
+class MemoryTableEntry:
+    def __init__(self, *, object_ref: dict, node_address: str, is_driver: bool,
+                 pid: int):
+        # worker info
+        self.is_driver = is_driver
+        self.pid = pid
+        self.node_address = node_address
+
+        # object info
+        self.object_size = int(object_ref.get("objectSize", -1))
+        self.call_site = object_ref.get("callSite", "<Unknown>")
+        self.object_id = ray.ObjectID(
+            decode_object_id_if_needed(object_ref["objectId"]))
+
+        # reference info
+        self.local_ref_count = int(object_ref.get("localRefCount", 0))
+        self.pinned_in_memory = bool(object_ref.get("pinnedInMemory", False))
+        self.submittedd_task_ref_count = int(
+            object_ref.get("submittedTaskRefCount", 0))
+        self.contained_in_owned = [
+            ray.ObjectID(decode_object_id_if_needed(object_id))
+            for object_id in object_ref.get("containedInOwned", [])
+        ]
+        self.reference_type = self._get_reference_type()
+
+    def is_valid(self) -> bool:
```

Comment:
In what cases is a memory entry not valid? Should this ever return false? Comment would be good.

Reply:
Added comment!